### PR TITLE
[MERGE] mail: allow the admin to open any emails

### PR DESCRIPTION
**Pull request by std-odoo**
When the admin opens an email, they might get an error message if they
can not access one of the attachment of the email or because of the ACL on
mail message.

In this merge we avoid that issue by fetching only accessible attachments
and allow checking related email_from, reply-to and subject as sudo.

Task-2464212

### CODE REVIEW

File: addons/mail/models/mail_message.py [hunk 1]
```diff
@@ -374,7 +374,7 @@ def _generate_model_record_ids(msg_val, msg_ids):
                     model_record_ids.setdefault(vals['model'], set()).add(vals['res_id'])
             return model_record_ids
 
-        if self.env.is_superuser():
+        if self.env.is_admin():
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:377]: Mmmh ... why this change ? Not sure we want this.

File: addons/mail/models/mail_mail.py [hunk 2]
```diff
@@ -81,6 +85,36 @@ def default_get(self, fields):
...
+    def _inverse_unrestricted_attachment_ids(self):
+        """We can only remove the attachments we have access to."""
+        for mail in self:
+            restricted_attaments = self.sudo().attachment_ids - self._filter_attachment_access(mail.attachment_ids)
+            mail.sudo().attachment_ids = restricted_attaments | mail.unrestricted_attachment_ids
+
+    def _filter_attachment_access(self, attachment_ids):
+        """Filter the given attachment to return only the records the current user have access to."""
+        ret_attachments = self.env['ir.attachment']
+        for attachment in attachment_ids:
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:108]: No way to do it in batch ? At least the "check_access_rights" ?

File: addons/mail/models/mail_mail.py [hunk 3]
```diff
@@ -81,6 +85,36 @@ def default_get(self, fields):
...
+        and the number of attachments we do not have access to.
+        """
+        for mail in self:
+            mail.unrestricted_attachment_ids = self._filter_attachment_access(mail.attachment_ids)
+            mail.restricted_attachment_count = len(mail.attachment_ids) - len(mail.unrestricted_attachment_ids)
+
+    def _inverse_unrestricted_attachment_ids(self):
+        """We can only remove the attachments we have access to."""
+        for mail in self:
+            restricted_attaments = self.sudo().attachment_ids - self._filter_attachment_access(mail.attachment_ids)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:102]: Using self in a loop ?

File: addons/mail/models/mail_mail.py [hunk 4]
```diff
@@ -81,6 +85,36 @@ def default_get(self, fields):
...
+            mail.sudo().attachment_ids = restricted_attaments | mail.unrestricted_attachment_ids
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:103]: Would sudo beforehand to sudo the whole recordset once

File: addons/mail/models/mail_mail.py [hunk 5]
```diff
@@ -81,6 +85,36 @@ def default_get(self, fields):
...
 
+    @api.depends('attachment_ids')
+    def _compute_restricted_attachments(self):
+        """We might not have access to all the attachments of the emails.
+
+        Compute the attachments we have access to,
+        and the number of attachments we do not have access to.
+        """
+        for mail in self:
+            mail.unrestricted_attachment_ids = self._filter_attachment_access(mail.attachment_ids)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:96]: Seems _filter_attachment_access could use mail as recordset, and use self.attachment_ids (therefore removing the need of a paraemter)

**Review by tde-banana-odoo — COMMENTED** [review]
Validation !

**Review by std-odoo — COMMENTED** [review]
@tde-banana-odoo Change done and answer about `self.env.is_admin` :)

File: addons/mail/models/mail_mail.py [hunk 6]
```diff
@@ -40,9 +41,16 @@ def default_get(self, fields):
 
     # content
     mail_message_id = fields.Many2one('mail.message', 'Message', required=True, ondelete='cascade', index=True, auto_join=True)
+
+    # Mail Message fields
+    message_int = fields.Integer(compute='_compute_message_int', compute_sudo=True)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:46]: ``mail_message_id_int`` maybe ? We already have that kind of fields (``mail_mail_id_int``, ``sms_sms_id_int``, ...) that way we keep the prefixes :)

File: addons/mail/models/mail_mail.py [hunk 7]
```diff
@@ -201,6 +245,21 @@ def _postprocess_sent_message(self, success_pids, failure_reason=False, failure_
...
+    def _add_inherited_fields(self):
+        """Allow to bypass ACLs for some mail message fields.
+
+        This trick add a related_sudo on the inherits fields, it can't be done with
+        >>> subject = fields.Char(related='mail_message_id.subject', related_sudo=True)
+        because the field of <mail.message> will be fetched two times (one time before of
+        the inherits, and a second time because of the related), and so it will add extra
+        SQL queries.
+        """
+        super()._add_inherited_fields()
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:258]: :exploding_head: is it somehow standard ?

File: addons/mail/models/mail_mail.py [hunk 8]
```diff
@@ -201,6 +245,21 @@ def _postprocess_sent_message(self, success_pids, failure_reason=False, failure_
             self.browse(mail_to_delete_ids).sudo().unlink()
         return True
 
+    @api.model
+    def _add_inherited_fields(self):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:249]: Linting: would define it after create/unlink, in the "CRUD" section (quite low level model :D )

File: addons/mail/models/mail_mail.py [hunk 9]
```diff
@@ -81,6 +89,42 @@ def default_get(self, fields):
...
+        """Filter the given attachment to return only the records the current user have access to."""
+        ret_attachments = self.env['ir.attachment']
+        if not self.check_access_rights('read', raise_exception=False):
+            return ret_attachments
+
+        for attachment in self.sudo().attachment_ids:
+            # Use SUDO here to not raise an error during the prefetch
+            # And then drop SUDO right to check if we can access it
+            try:
+                attachment.sudo(False).check('read')
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:122]: Shame that ``check`` does not support working like ``_filter_access_rules`` who keep only valid records ... I suppose you have to do a per-record basis because it raises.

Do you think this would be much work to have it available on attachment model (would mean somehow re-implementing the check / extracting common parts, ...). Asking for a friend :)

File: addons/mail/models/mail_mail.py [hunk 10]
```diff
@@ -40,9 +41,16 @@ def default_get(self, fields):
...
     mail_message_id = fields.Many2one('mail.message', 'Message', required=True, ondelete='cascade', index=True, auto_join=True)
+
+    # Mail Message fields
+    message_int = fields.Integer(compute='_compute_message_int', compute_sudo=True)
+
     body_html = fields.Text('Rich-text Contents', help="Rich-text/HTML message")
     references = fields.Text('References', help='Message references, such as identifiers of previous messages', readonly=1)
     headers = fields.Text('Headers', copy=False)
+    restricted_attachment_count = fields.Integer('Restricted attachments', compute='_compute_restricted_attachments')
+    unrestricted_attachment_ids = fields.Many2many('ir.attachment', string='Unrestricted Attachments',
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:52]: Would completely love having an unit test for that (especially with the inverse as this is nice, but definitively prone to be broken someday :D ) .

**Review by tde-banana-odoo — COMMENTED** [review]
Riviou: some comments, some proposals ! We are moving forward, step after step. Courage !

File: addons/mail/models/mail_mail.py (hunk 9 above)
Comment (std-odoo) [reply addons/mail/models/mail_mail.py]: it's a bit tricky because you read read acl to execute the method on the records (and if we execute in SUDO, we should see all attachment, etc)

but if it can be useful, what I can do is a model method, call it with the attachments ids in arguments

what do you think ?

**Review by std-odoo — COMMENTED** [review]
@tde-banana-odoo Change done and test added :)

File: addons/mail/models/mail_mail.py [hunk 11]
```diff
@@ -40,9 +41,16 @@ def default_get(self, fields):
 
     # content
     mail_message_id = fields.Many2one('mail.message', 'Message', required=True, ondelete='cascade', index=True, auto_join=True)
+
+    # Mail Message fields
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:45]: Probably unnecessary comment by the way ^^ field name is clear enough to be put just after mail_message_id

File: addons/test_mail/tests/test_mail_mail.py [hunk 12]
```diff
@@ -60,6 +61,61 @@ def _reset_data(self):
...
+            mail.write({
+                'unrestricted_attachment_ids': [Command.link(new_attachment.id)],
+            })
+            self.assertEqual(mail.restricted_attachment_count, 2)
+            self.assertEqual(len(mail.unrestricted_attachment_ids), 3)
+            self.assertEqual(mail.unrestricted_attachment_ids.mapped('name'), ['file 1', 'file 3', 'new file'])
+            self.assertEqual(len(mail.attachment_ids), 5)
+
+            # Remove an attachment
+            mail.write({
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_mail.py:107]: We could try to reset command ((5, 0) and check other attachments are still there :)

**Review by tde-banana-odoo — COMMENTED** [review]
LGTM, one small remark and one test proposal :)

File: addons/test_mail/tests/test_mail_mail.py (hunk 12 above)
Comment (std-odoo) [reply addons/test_mail/tests/test_mail_mail.py:107]: good idea :)

File: addons/mail/models/mail_mail.py [hunk 13]
```diff
@@ -40,9 +40,15 @@ def default_get(self, fields):
 
     # content
     mail_message_id = fields.Many2one('mail.message', 'Message', required=True, ondelete='cascade', index=True, auto_join=True)
+
+    mail_message_id_int = fields.Integer(compute='_compute_mail_message_id_int', compute_sudo=True)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_mail.py:44]: Could you remove lines between fields definition, unless there is a reason to add lines like that ? If you want to order fields, why not a comment per "section or category of fields", but those void lines feel weird to me ...

Comment (std-odoo) [reply addons/mail/models/mail_mail.py:43]: @tde-banana-odoo done :)

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
